Proposal of Fractionalized NFT Standard (#3525)
ethereum/EIPs: EIPS/eip-3525.md

---
eip: 3525
title: Fractionalized NFT Standard

> author: Will Wang <@will-edge>, Mike Meng <myan@solv.finance>, TsaiYee <yee.tsai@gmail.com>, Ryan Chow <ryanchow@solv.finance>, Zhongxin Wu <wuzhongxin@solv.finance>, AlvisDu <dujincai@solv.finance>

Reviewer: ```suggestion
author: Will Wang (@will-edge), Mike Meng <myan@solv.finance>, TsaiYee <yee.tsai@gmail.com>, Ryan Chow <ryanchow@solv.finance>, Zhongxin Wu <wuzhongxin@solv.finance>, AlvisDu <dujincai@solv.finance>
```
this is the proper syntax

### Approval

For maximum semantical compatibility with ERC-721, as well as simplifying the approval model, we decided to make the relationship between two levels of approval like that:

1. Approval of an id does not result in the ability to partial transfer units from this id by the approved operator;
2. Approval of all units in a token does not result in the ability to transfer the token entity by the approved operator;
3. `setApprovalForAll` will result in the ability to transfer any tokens from the owner, as well as the ability to partial transfer units from any token.
3. `setApprovalForAll` will result in the ability to approve any tokens of the owner to third parties, as well as the ability to approve partial transfer units of any token to third parties.

## Backwards Compatibility
	
As mentioned at the very beginning, a VNFT contract is basically an ERC721 contract, hence it is 100% compatible with ERC-721.

> ## References

Reviewer: Please remove this section.

Author: OK, got it. We had checked some old EIPs and saw some 'References' section, and treated it as a required section, didn't noticed EIP-1's newest structure.

---
eip: <to be assigned>

> title: <Versatile NFT Standard>

Reviewer: It seems like a title like "Fractionalized NFT Standard" would more accurately represent the ERC you've proposed.

Author: Thanks for the suggestion, we had discussed many names for this standard include 'Fractionalized' one, but now seems this one is easier for understand. Now changed it to "Proposal for Fractionalized NFT Standard", however temporally left VFNT symbol unchanged for ease of modifications, once this one is accepted more, may think of changing the symbol to FNFT or sth. like that.

> author: will-edge <will@unizon.pro>, Mike Meng <myan@solv.finance>, TsaiYee <yee.tsai@gmail.com>, Ryan Chow <ryanchow@solv.finance>, Zhongxin Wu <wuzhongxin@solv.finance>, AlvisDu <dujincai@solv.finance>

Reviewer: Still waiting on this.

	
The most simple and direct way to solve the problem is to add an amount property directly to ERC-721 token, making it best for both property customization and semi-fungibility. Nevertheless, the ERC-721 compatibility will help the new standard easily utilizing existing infrastructures, resulting in fast adoptance.

	

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

**Every contract implementing the VNFT standard MUST implement the `VNFT` interfaces as follows:**

```solidity
pragma solidity 0.7.6;

/**
 * @title VNFT Versatile Non-Fungible Token Standard
 * @dev See https://eips.ethereum.org/EIPS/eip-VNFT
 * Note: the ERC-165 identifier for this interface is 0x1487d183.
 */

> interface VNFT /* is ERC721 */{

Reviewer: ```suggestion
interface VNFT /* is ERC-721 */{
```

        This function MUST revert if it rejects the transfer.
        Return of any other value than the prescribed keccak256 generated value MUST result in the transaction being reverted by the caller.
        @param operator  The address which initiated the transfer (i.e. msg.sender)
        @param from      The address which previously owned the token
        @param id        The ID of the token being transferred
        @param units     The units of tokenId being transferred
        @param data      Additional data with no specified format
        @return           `bytes4(keccak256("onVNFTReceived(address,address,uint256,uint256,bytes)"))`
	Note: the ERC-165 identifier for this interface is 0xb382cdcd.
 */
interface IVNFTReceiver {
    function onVNFTReceived(address operator, address from, uint256 tokenId,
        uint256 units, bytes calldata data) external returns (bytes4);
}
```

### Token Manipulation

#### Scenarios

**_Transfer:_**
	
Since a VNFT token is ERC-721 compatible, it has ID level transfer and units level transfer.

> The ID level transfer SHOULD obey ERC721 transfer rules with neither extra restrictions nor special treatments.

Reviewer: ```suggestion
The ID level transfer SHOULD obey ERC-721 transfer rules with neither extra restrictions nor special treatments.
```

	- MUST revert if either `_tokenId` or `_targetTokenId` is not a valid token.
	- MUST revert if `_from` is not the current owner of `_tokenId`.
	- MUST revert if `_to` is not the current owner of `_targetTokenId`.
	- MUST revert if `_to` is the zero address.
	- MUST revert if the transfer amount exceeds the actual amount of units in `_tokenId`.
	- MUST revert if the transfer amount exceeds the VNFT approved units limit.
	- MUST emit the VNFT level `PartialTransfer` event.

**_safeTransferFrom rules:_**

- `safeTransferFrom` SHOULD be used to implement the same function as `transferFrom`, with an extra step to check if the recipient is capable of receiving VNFTs by implementing the `onVNFTReceived` interface.
- MUST obey the above rules set for `transferFrom`.
- MUST check if `_to` is a smart contract (code size > 0). If so, `safeTransferFrom` MUST call `onVNFTReceived` on `_to` and MUST revert if the return value does not match `bytes4(keccak256("onVNFTReceived(address,address,uint256,uint256,bytes)"))`

### Metadata 

#### Metadata Extensions

> VNFT metadata extensions are compatible ERC721 metadata extensions.

Reviewer: ```suggestion
VNFT metadata extensions are compatible ERC-721 metadata extensions.
```

        },
         "decimals": {
            "type": "integer",
            "description": "The number of decimal places that the token amount should display - e.g. 18, means to divide the token amount by 1000000000000000000 to get its user representation."
        },
        "properties": {
            "type": "object",
            "description": "Arbitrary properties. Values may be strings, numbers, object or arrays."
        }
    }
}
```

### Approval

VNFT adds a new approval model, that is, one can approve operators to partial transfer units from a token with certain ID, the new interface is:
```function approve(address to, uint256 tokenId, uint256 units); ```

## Rationale 

### Metadata generation

Since VNFT is designed for representing underlying assets, rather than artifacts for gaming or arts, the implementation should give out the metadata directly from contract code, rather than give a URL of a server for returning metadata.

### Design decision: Keep unsafe transfer

There are mainly two reasons we keep the unsafe transfer interfaces:

> 1. Since VNFT is ERC721 compatible, we must keep compatibility for all wallets and contracts that are still calling unsafe transfer interfaces for ERC721 tokens.

Reviewer: ```suggestion
1. Since VNFT is ERC-721 compatible, we must keep compatibility for all wallets and contracts that are still calling unsafe transfer interfaces for ERC-721 tokens.
```

2. We want to keep the ability that dapps can trigger business logic on contracts by simply transferring VNFT tokens to them, that is, a contract can put business logic in `onVNFTReceived` function so that it can be called whenever a token is transferred using `safeTransferFrom`. However, in this situation, an approved contract with customized transfer functions like deposit etc. SHOULD never call `safeTransferFrom` since it will result in confusion that whether `onVNFTReceived` is called by itself or other dapps that safe transfer a token to it.

### Approval

> For maximum semantical compatibility with ERC721, as well as simplifying the approval model, we decided to make the relationship between two levels of approval like that:

Reviewer: ```suggestion
For maximum semantical compatibility with ERC-721, as well as simplifying the approval model, we decided to make the relationship between two levels of approval like that:
```

1. Approval of an id does not result in the ability to partial transfer units from this id by the approved operator;
2. Approval of all units in a token does not result in the ability to transfer the token entity by the approved operator;
3. `setApprovalForAll` will result in the ability to transfer any tokens from the owner, as well as the ability to partial transfer units from any token.
3. `setApprovalForAll` will result in the ability to approve any tokens of the owner to third parties, as well as the ability to approve partial transfer units of any token to third parties.

## Backwards Compatibility
	

> As mentioned at the very beginning, a VNFT contract is basically an ERC721 contract, hence it is 100% compatible with ERC721.

Reviewer: ```suggestion
As mentioned at the very beginning, a VNFT contract is basically an ERC721 contract, hence it is 100% compatible with ERC-721.
```

> ## References

Reviewer: This isn't a valid top level section of an ERC per [EIP-1](https://eips.ethereum.org/EIPS/eip-1), please remove this section.

Reviewer: Still waiting on this too.

Author: Hi, thanks for the reminding, we had correct the names/emails now, but do you mean References is not a top level section of ERC? I saw from EIP-1 that there is a References section, not sure wether there is some misunderstanding on this, thanks in advance.

Reviewer: @will-edge I think you are confusing "Reference" and "Reference Implementation". There is no "Reference" section in EIPs.

**Standards**
- [ERC-721 Non-Fungible Token Standard](./eip-721.md)
- [ERC-165 Standard Interface Detection](./eip-165.md)
- [JSON Schema](https://json-schema.org/)
- [RFC 2119 Key words for use in RFCs to Indicate Requirement Levels](https://www.ietf.org/rfc/rfc2119.txt)

> **Implementations**

Reviewer: This can go under the "Reference Implementation" section, per EIP-1.
